walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,93 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> results = new ArrayList<>();
+    Collection<String> allAttendees = request.getAttendees();
+    int numAttendees = allAttendees.size();
+    long meetingTime = request.getDuration();
+
+    // Base Case: No required attendees.
+    if (numAttendees == 0) {
+        results = Arrays.asList(TimeRange.WHOLE_DAY);
+        return results;
+    // Base Case: Invalid meeting time.
+    } else if (meetingTime == 0 || meetingTime > TimeRange.WHOLE_DAY.duration()) {
+        results = Arrays.asList();
+        return results;
+    }
+
+    // Extract and sort event times (i.e. busy times) by start time.
+    List<TimeRange> timeRanges = new ArrayList<>();
+    for (Event e : events) {
+        // Ignore event times of people not actually attending.
+        boolean isOptional = isEventOptional(e, allAttendees);
+        if (!isOptional) {
+            timeRanges.add(e.getWhen());
+        }
+    }
+    timeRanges.sort(TimeRange.ORDER_BY_START);
+
+    // Track latest end time of any event.
+    // Since timeRanges was sorted by start, the event with the latest end time
+    // might come before the last event in the List (e.g. if we had [5, 20]
+    // followed by [10, 15], we want to use [20, END_OF_DAY] for the final gap,
+    // not [15, END_OF_DAY]).
+    int endOfLatestEvent = 0;
+    TimeRange gap;
+
+    // Iteratively construct gaps between busy times.
+    for (int i = 0; i < timeRanges.size(); i++) {        
+        // First gap before the first time anyone is busy.
+        if (i == 0) {
+            gap = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, timeRanges.get(i).start(), false);
+        // gaps in the middle.
+        } else {
+            gap = TimeRange.fromStartEnd(timeRanges.get(i-1).end(), timeRanges.get(i).start(), false);
+        }
+
+        // Prevent edge behavior since start times are monotonically increasing,
+        // but end times are not.
```

Comment:
I think you can still get into trouble here.
But let's see if a test can catch this if you make it a bit more complicated than the multi-event overlap tests that the code starts with.

Reply:
I thought up some edge cases where these gaps still might not work out, so I agree that the check should be a bit more complicated.